usr/src/uts/common/io/scsi/adapters/smartpqi/smartpqi_util.c

```diff
@@ -0,0 +1,947 @@
+	case PQI_EVENT_TYPE_AIO_CONFIG_CHANGE:
+		return ("AIO Config Change");
+	case PQI_EVENT_TYPE_HEARTBEAT: return ("Heartbeat");
+	default: return ("Unsupported Event Type");
+	}
+}
+
+char *
+bool_to_str(int v)
+{
+	return (v ? "T" : "f");
+}
+
+char *
+dtype_to_str(int t)
+{
+	switch (t) {
+	case DTYPE_DIRECT: return ("Direct");
+	case DTYPE_SEQUENTIAL: return ("Sequential");
+	case DTYPE_ESI: return ("ESI");
+	case DTYPE_ARRAY_CTRL: return ("RAID");
+	default: return ("Ughknown");
+	}
+}
+
+static ddi_dma_attr_t single_dma_attrs = {
+	DMA_ATTR_V0,	/* attribute layout version			*/
+	0x0ull,		/* address low - should be 0 (longlong)		*/
+	0xffffffffffffffffull, /* address high - 64-bit max		*/
+	0x7ffffull,	/* count max - max DMA object size		*/
+	4096,		/* allocation alignment requirements		*/
+	0x78,		/* burstsizes - binary encoded values		*/
+	1,		/* minxfer - gran. of DMA engine		*/
+	0x007ffffull,	/* maxxfer - gran. of DMA engine		*/
+	0xffffffffull,	/* max segment size (DMA boundary)		*/
+	1,		/* For pqi_alloc_single must be contig memory	*/
+	512,		/* granularity - device transfer size		*/
+	0		/* flags, set to 0				*/
+};
+
+pqi_dma_overhead_t *
+pqi_alloc_single(pqi_state_t s, size_t len)
+{
+	pqi_dma_overhead_t	*d;
+	ddi_dma_cookie_t	cookie;
+
+	d = kmem_zalloc(sizeof (*d), KM_SLEEP);
+	d->len_to_alloc = len;
+
+	if (ddi_dma_alloc_handle(s->s_dip, &single_dma_attrs,
+	    DDI_DMA_SLEEP, 0, &d->handle) != DDI_SUCCESS)
+		goto error_out;
+
+	if (ddi_dma_mem_alloc(d->handle, len, &s->s_reg_acc_attr,
+	    DDI_DMA_CONSISTENT, DDI_DMA_SLEEP, 0,
+	    &d->alloc_memory, &len, &d->acc) != DDI_SUCCESS)
+		goto error_out;
+
+	(void) memset(d->alloc_memory, 0, len);
+	if (ddi_dma_addr_bind_handle(d->handle, NULL, d->alloc_memory, len,
+	    DDI_DMA_RDWR, DDI_DMA_SLEEP, 0, &cookie, &d->cookie_count) !=
+	    DDI_SUCCESS)
+		goto error_out;
+
+	d->dma_addr = cookie.dmac_laddress;
+	if (d->cookie_count != 1)
+		ddi_dma_nextcookie(d->handle, &d->second);
+
+	return (d);
+
+error_out:
+	pqi_free_single(s, d);
+	return (NULL);
+}
+
+void
+pqi_free_single(pqi_state_t s, pqi_dma_overhead_t *d)
+{
+	(void) ddi_dma_unbind_handle(d->handle);
+	if (d->alloc_memory != NULL)
+		ddi_dma_mem_free(&d->acc);
+	if (d->handle != NULL)
+		ddi_dma_free_handle(&d->handle);
+	ASSERT(s->s_dip != NULL);
+	kmem_free(d, sizeof (*d));
+}
+
+void
+pqi_show_dev_state(pqi_state_t s)
+{
+	uint32_t dev_status = G32(s, pqi_registers.device_status);
+
+	switch (dev_status & 0xf) {
+	case 0:
+		cmn_err(CE_NOTE, "Power_On_And_Reset\n");
+		break;
+
+	case 1:
+		cmn_err(CE_NOTE, "PQI_Status_Available\n");
+		break;
+
+	case 2:
+		cmn_err(CE_NOTE, "All_Registers_Ready\n");
+		break;
+
+	case 3:
+		cmn_err(CE_NOTE,
+		    "Adminstrator_Queue_Pair_Ready\n");
+		break;
+
+	case 4:
+		cmn_err(CE_NOTE, "Error: %s %s\n",
+		    dev_status & 0x100 ? "(OP OQ Error)" : "",
+		    dev_status & 0x200 ? "(OP IQ Error)" : "");
+		show_error_detail(s);
+		break;
+	}
+}
+
+void *
```

Comment:
Why do we need a separate allocation function here?  kmem debugging is pretty darned good, so I don't understand why we need another form of debugging allocator here.